### PR TITLE
Enable matrix_load, matrix_store, and matrix_mad

### PR DESCRIPTION
**Pull request by yubingex007-a11y**
Change-Id: I97b939c5559317ad82bec460221bccb83975b004

### CODE REVIEW

**Review by bader — CHANGES_REQUESTED** [review]
Please, put files to `ext/intel` directory instead of `ONEAPI`.

**Comment by dkhaldi** [comment]
> Please, put files to `ext/intel` directory instead of `ONEAPI`.

@bader, Although the matrix feature is in experimental phase, it is a oneAPI feature, it is not an intel specific feature. I think we should keep it as it was already.

**Comment by bader** [comment]
Sorry, I meant to say `ext/oneapi`. Please, rebase on the tip of the branch to resolve the conflicts.
See d703f578fcf7e8fb4716846bdd115add128a345b for more details on the requested change. We moved all oneAPI extensions from `ONEAPI` to `ext/oneapi` directory.

**Comment by dkhaldi** [comment]
> Sorry, I meant to say `ext/oneapi`. Please, rebase on the tip of the branch to resolve the conflicts.
> See [d703f57](https://github.com/intel/llvm/commit/d703f578fcf7e8fb4716846bdd115add128a345b) for more details on the requested change. We moved all oneAPI extensions from `ONEAPI` to `ext/oneapi` directory.

What would be the namespace we should use in this case? 
Currently we are using: sycl:ext::intel::experimental::matrix;
Should it be: sycl:ext::ONEAPI::experimental::matrix;?
BTW, I see experimental directory in ONEAPI/. Should we move matrix/ there?

**Comment by romanovvlad** [comment]
> 
> 
> The namespace should be `sycl::ext::oneapi::experimental::matrix`.
> 
> According to my understanding `ONEAPI` directory is left for backward compatibility only and should not be used for new features.
> All new features should follow SYCL spec guidelines and use `sycl::ext::oneapi` namespace (implementation guideline is to put files into `ext/oneapi` directory reflecting namespace structure).
> 
> Tagging @dm-vodopyanov and @romanovvlad to correct me if I miss something.

Please, use `sycl::ext::oneapi::experimental` for general extension and `sycl::ext::intel::experimental` if the extension is specific to Intel HW.

File: sycl/include/sycl/ext/oneapi/matrix/matrix-aot-amx.hpp
Comment (yubingex007-a11y) [inline sycl/include/sycl/ext/oneapi/matrix/matrix-aot-amx.hpp:41]: BTW, @bader @dkhaldi , should we move this file to ext/intel directory since it works on intel HW?

Comment (yubingex007-a11y) [reply sycl/include/sycl/ext/oneapi/matrix/matrix-aot-amx.hpp:41]: mistakenly change it, sorry for inconvience.

File: sycl/include/sycl/ext/oneapi/matrix/matrix-jit.hpp
Comment (AlexeySotkin) [inline sycl/include/sycl/ext/oneapi/matrix/matrix-jit.hpp:64]: Template parameter `L` is not used. probably we should do 
```suggestion
      Ptr, stride, spv_matrix_layout_traits<L>::value);
```

File: sycl/include/sycl/ext/oneapi/matrix/matrix-jit.hpp
Comment (AlexeySotkin) [inline sycl/include/sycl/ext/oneapi/matrix/matrix-jit.hpp:80]:
```suggestion
    Group sg, joint_matrix<Group, T, NumRows, NumCols, Layout> &obj,
```

File: sycl/include/sycl/ext/oneapi/matrix/matrix-jit.hpp
Comment (AlexeySotkin) [inline sycl/include/sycl/ext/oneapi/matrix/matrix-jit.hpp:86]:
```suggestion
      Ptr, res.spvm, stride, spv_matrix_layout_traits<L>::value);
```

File: sycl/include/sycl/ext/oneapi/matrix/matrix-jit.hpp
Comment (AlexeySotkin) [inline sycl/include/sycl/ext/oneapi/matrix/matrix-jit.hpp:59]: Function argument `L` is not used. Probably we should use it as operand of `__spirv_MatrixLoadINTEL`, but it must be a constant.
It seems like we have two layouts here: one for the memory we read from, and another for the matrix object, right? If so, I think we need another template parameter to specify layout of the memory.

File: sycl/include/sycl/ext/oneapi/matrix/matrix-aot-amx.hpp
Comment (dkhaldi) [reply sycl/include/sycl/ext/oneapi/matrix/matrix-aot-amx.hpp]: > BTW, @bader @dkhaldi , should we move this file to ext/intel directory since it works on intel HW?

Please keep everything in ext/oneapi.
The feature is still oneAPI "matrix". It is just that the initial implementation targeted only AMX.

File: sycl/include/sycl/ext/oneapi/matrix/matrix-jit.hpp
Comment (dkhaldi) [reply sycl/include/sycl/ext/oneapi/matrix/matrix-jit.hpp:59]: __spirv_MatrixLoadINTEL  already takes layout as argument and this should be L, the memory layout.
Are we using the matrix layout for that argument? If yes this should be changed.

The matrix layout is kept on the matrix type itself.

File: sycl/include/sycl/ext/oneapi/matrix/matrix-jit.hpp
Comment (yubingex007-a11y) [inline sycl/include/sycl/ext/oneapi/matrix/matrix-jit.hpp:64]: L

File: sycl/include/sycl/ext/oneapi/matrix/matrix-jit.hpp
Comment (yubingex007-a11y) [inline sycl/include/sycl/ext/oneapi/matrix/matrix-jit.hpp:55]: MatrixLayout

File: sycl/test/matrix/matrix-amx-bf16-test.cpp
Comment (dkhaldi) [inline sycl/test/matrix/matrix-amx-bf16-test.cpp:3]: // Only the leader perform AMX computation.
Why do you have "the leader" test  in the code. this is wrong for two reasons:
1- SG size is one so we don't need to test that.
2- all work items in the subgroup must enter the joint matrix code. So we should have diverge code in there

File: sycl/test/matrix/matrix-bf16-test-SG-16.cpp
Comment (dkhaldi) [inline sycl/test/matrix/matrix-bf16-test-SG-16.cpp:71]: Again, we should not have "the leader" test in the code. this is wrong because all work items in the subgroup must enter the joint matrix code. So we should not have diverge code in there. 
The OpenCL backend will not vectorize the code so we should be able to generate ONE AMX code for the whole sub group

File: sycl/test/matrix/matrix-bf16-test.cpp
Comment (dkhaldi) [inline sycl/test/matrix/matrix-bf16-test.cpp:71]: same comment as above

File: sycl/test/matrix/matrix-int8-test-SG-16.cpp
Comment (dkhaldi) [inline sycl/test/matrix/matrix-int8-test-SG-16.cpp:71]: same comment as above

File: sycl/test/matrix/matrix-int8-test.cpp
Comment (dkhaldi) [inline sycl/test/matrix/matrix-int8-test.cpp:71]: same comment as above

File: sycl/test/matrix/matrix-int8-test.cpp
Comment (dkhaldi) [inline sycl/test/matrix/matrix-int8-test.cpp:73]: We should set the stride to N right? the user should not worry about the VNNI factor in general

File: sycl/include/sycl/ext/oneapi/matrix/matrix-jit.hpp
Comment (AlexeySotkin) [reply sycl/include/sycl/ext/oneapi/matrix/matrix-jit.hpp:59]: IMO `memL` and `matL` are not quite user friendly names. Are `memory_layout` and `matrix_layout` more readable?

Comment (dkhaldi) [reply sycl/include/sycl/ext/oneapi/matrix/matrix-jit.hpp:59]: Right now, the user does not need to specify any template argument when calling load/store. If we move the memory layout to the template, the user will have to specify each of the parameters (deduce all or nothing) when calling load/store. 

The other reason is that I would like to not converge too much (unless there is a big reason) from Vulkan and WMMA APIs thus keeping the meory layout as argument to the functions

File: sycl/include/sycl/ext/oneapi/matrix/matrix-jit.hpp
Comment (yubingex007-a11y) [reply sycl/include/sycl/ext/oneapi/matrix/matrix-jit.hpp:80]: Do we need to change joint_matrix_load as well?

File: sycl/include/sycl/ext/oneapi/matrix/matrix-jit.hpp
Comment (yubingex007-a11y) [reply sycl/include/sycl/ext/oneapi/matrix/matrix-jit.hpp:59]: I guess MemL is behind MatL in Template parameters list?

File: sycl/include/sycl/ext/oneapi/matrix/matrix-jit.hpp
Comment (AlexeySotkin) [reply sycl/include/sycl/ext/oneapi/matrix/matrix-jit.hpp]: Just improving the name. There is no result for the store operation. It returns void. This parameter is a matrix _object_ which we want to store.

File: sycl/include/sycl/ext/oneapi/matrix/matrix-jit.hpp
Comment (dkhaldi) [inline sycl/include/sycl/ext/oneapi/matrix/matrix-jit.hpp:69]: you omitted the group argument here right?

File: sycl/include/sycl/ext/oneapi/matrix/matrix-jit.hpp
Comment (dkhaldi) [inline sycl/include/sycl/ext/oneapi/matrix/matrix-jit.hpp:76]: same as above

File: sycl/include/sycl/ext/oneapi/matrix/matrix-jit.hpp
Comment (dkhaldi) [inline sycl/include/sycl/ext/oneapi/matrix/matrix-jit.hpp:82]: spirv should have scope argument so sg should be passed in all load/store/mad functions

File: sycl/include/sycl/ext/oneapi/matrix/matrix-jit.hpp
Comment (dkhaldi) [inline sycl/include/sycl/ext/oneapi/matrix/matrix-jit.hpp:122]: I thought we said obj should be change to "src" right?

**Review by dkhaldi — CHANGES_REQUESTED** [review]

**Comment by dkhaldi** [comment]
LGTM

**Comment by bader** [comment]
> LGTM

@dkhaldi, you must approve using GitHub UI to unblock the merge. Currently it's blocked by your previous review where changes were requested.

**Review by dkhaldi — APPROVED** [review]

**Comment by yubingex007-a11y** [comment]
@intel/llvm-reviewers-runtime Hi, Could you do a code review please?

File: sycl/include/sycl/ext/oneapi/matrix/matrix-jit.hpp
Comment (dkhaldi) [inline sycl/include/sycl/ext/oneapi/matrix/matrix-jit.hpp:47]: Bing, sorry for the last minute review but I got two requests from the design meeting today:
1- we should move the Group template argument on joint_matrix to the last position and make it optional. By default set Group to subgroup. So the user does not need to specify this each time
2- change the "mad" function name from joint_matrix_mad to joint_matrix_mma.
Can you make these changes before merge?

**Review by dkhaldi — CHANGES_REQUESTED** [review]
Bing, sorry for the

**Review by dkhaldi — APPROVED** [review]

**Comment by dkhaldi** [comment]
LGTM, @bader can you help merge this PR?

**Comment by bader** [comment]
> LGTM, @bader can you help merge this PR?

The PR is waiting on code owner review from @intel/llvm-reviewers-runtime.

**Review by s-kanaev — COMMENTED** [review]
`matrix-amx.hpp => matrix-aot-amx.hpp`
Does this renaming follow API/ABI compatibility policy?

`sycl/include/CL/sycl/ONEAPI/matrix/matrix-amx.hpp` said one should include `matrix-amx.hpp` and the latter file becomes unavailable now.

Probably, the user won't be happy.

Other than that, the changes seem to be legit.

**Comment by yubingex007-a11y** [comment]
> `matrix-amx.hpp => matrix-aot-amx.hpp`
> Does this renaming follow API/ABI compatibility policy?
> 
> `sycl/include/CL/sycl/ONEAPI/matrix/matrix-amx.hpp` said one should include `matrix-amx.hpp` and the latter file becomes unavailable now.
> 
> Probably, the user won't be happy.
> 
> Other than that, the changes seem to be legit.
@s-kanaev , Thanks for your review.
This is an experimental interface so it is assumed to be changed.

**Review by s-kanaev — APPROVED** [review]
LGTM

**Pull request merged**
